src/main/scala/org/ergoplatform/nodeView/wallet/ErgoWalletActor.scala

```diff
@@ -96,19 +98,48 @@ class ErgoWalletActor(settings: ErgoSettings,
     }
   }
 
+  def scanBlock(block: ErgoFullBlock): Unit = {
+    log.info(s"Wallet is going to scan a block ${block.id} at height ${block.height}")
+    val (reg, offReg) =
+      WalletScanLogic.scanBlockTransactions(registry, offChainRegistry, stateContext, walletVars, block)
+    registry = reg
+    offChainRegistry = offReg
+  }
+
   private def scanLogic: Receive = {
     //scan mempool transaction
     case ScanOffChain(tx) =>
       val newWalletBoxes = WalletScanLogic.extractWalletOutputs(tx, None, walletVars)
       val inputs = WalletScanLogic.extractInputBoxes(tx)
       offChainRegistry = offChainRegistry.updateOnTransaction(newWalletBoxes, inputs)
 
+    case ScanInThePast(blockHeight) =>
+      val lastHeight = registry.fetchDigest().height
+      val expectedHeight = lastHeight + 1
+      if (expectedHeight == blockHeight) {
+        historyReader.bestFullBlockAt(blockHeight) match {
+          case Some(block) =>
+            scanBlock(block)
+          case None =>
+          // We may do not have a block if, for example, the blockchain is pruned. This is okay, just skip it.
+        }
+        if (blockHeight < fullHeight) {
+          self ! ScanInThePast(blockHeight + 1)
+        }
+      }
+
     //scan block transactions
     case ScanOnChain(block) =>
-      val (reg, offReg) =
-        WalletScanLogic.scanBlockTransactions(registry, offChainRegistry, stateContext, walletVars, block)
-      registry = reg
-      offChainRegistry = offReg
+      val lastHeight = registry.fetchDigest().height
+      val expectedHeight = lastHeight + 1
+      if (expectedHeight == block.height) {
+        scanBlock(block)
+      } else if (expectedHeight < block.height) {
+        log.warn(s"Wallet: skipped blocks found starting from $expectedHeight, going back to scan them")
+        self ! ScanInThePast(expectedHeight)
+      } else {
+        log.warn(s"Wallet: $expectedHeight > block.height for a block at ${block.height}, blockId: ${block.id}")
```

Comment:
What this message means?

Reply:
This message means wallet is in some strange state (got the block in the past to scan), thus the developers need to investigate what can be a reason for that.

Reply:
@kushti It doesn't make much sense to add clarifications here in conversations without changing the code itself. Any question actually means the code is not clear enough or/and require improvement (i.e. add comments). So when I ask the question about the meaning of the code (which I do from time to time) I usually expect to see the answer in the code.

Reply:
I think the answer is clear enough, the message provides devs heights and block id , along with other lines in the log it can be hopefully useful for investigations.

Reply:
This message in the log is cryptic. That is my point. It is better to give more descriptive message in the log, at least to easily search it in the log file.

Reply:
Well, let me explain how the logs are being used (based on tens of cases for the first years of the mainnet existence). A user is sending logs (to me, as only me is doing support atm unfortunately), then I am looking ERROR messages, then WARN, then I look into the message to extract details (like ids) to find correlated messages.

In regards with the workflow, the message looks good to me. If you can propose improvements, please do. To be honest, I do not understand your intentions and concerns.